nucypher/blockchain/eth/sol/source/contracts/staking_contracts/PreallocationEscrow.sol

```diff
@@ -62,7 +73,18 @@ contract PreallocationEscrow is AbstractStakingContract, Ownable {
     * @param _value Amount of token to withdraw
     **/
     function withdrawTokens(uint256 _value) public onlyOwner {
-        require(token.balanceOf(address(this)).sub(getLockedTokens()) >= _value);
+        uint256 balance = token.balanceOf(address(this));
+        uint32 secondsPerPeriod = stakingEscrow.secondsPerPeriod();
+        uint16 currentPeriod = uint16(block.timestamp / secondsPerPeriod);
+        uint16 endLockPeriod = uint16(endLockTimestamp / secondsPerPeriod);
+        if (currentPeriod <= endLockPeriod) {
+            uint256 stakedTokens = stakingEscrow.getLockedTokens(address(this), endLockPeriod - currentPeriod);
```

Comment:
Maybe it's too strict, because only counted tokens that locked for whole period. How about `getAllTokens` function? it will return all tokens that belong to current `PreallocationEscrow` and will simplify this logic

Reply:
In the beginning `x` tokens belong to  `PreallocationEscrow` (PE). 
Next step is staking (transfer+lock), `x = y + z`, `y` -> PE, `z` -> locked in `StakingEscrow` (SE), `balance = y`, `stakedTokens = z`. 
After some time we get reward or some of `z` amount become unlocked, `x = y + t + v + w`, `y` -> PE, `t` -> locked in SE, `v` -> reward in SE, `w` -> unlocked in SE (we can track only sum `v + w`). Still `balance = y`, `stakedTokens = t` but `v + w` ignored. While this not locked tokens in SE they are not taken in calculation for PE, they become withdrawable only after transfering back . I think it's to strict and complex. We can use all amount `t + v + w` just calling `getAllTokens`, all this tokens belong to PE and can't be withdrawn except through PE so this's safe. Also this method much simplier then `getLockedTokens` (just returns `stakerInfo.value`) and not need to think about periods.